WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,605 @@
+     * - Parameter stream: name of the stream
+     * - Returns: `true` if in sample or `false` otherwise
+     *
+     * The determinations are lazy and cached, so each stream's in-sample vs out-of-sample determination
+     * is computed only once, the first time an event is logged to that stream.
+     *
+     * Refer to sampling settings section in
+     * [mw:Wikimedia Product/Analytics Infrastructure/Stream configuration](https://www.mediawiki.org/wiki/Wikimedia_Product/Analytics_Infrastructure/Stream_configuration)
+     * for more information.
+     */
+    private func in_sample(stream: String) -> Bool {
+
+        if !cache.keys.contains(stream) {
+            let stream_entry: [String: Any] = CONFIG![stream]!
+            if stream_entry.keys.contains("config") {
+                let stream_config = stream_entry["config"] as! [String: Any]
+                if stream_config.keys.contains("sampling") {
+                    let sampling_config = stream_config["sampling"] as! [String: Any]
+                    /*
+                     * If cache does not have a determination for stream,
+                     * generate one.
+                     */
+                    if !sampling_config.keys.contains("rate") {
+                        /*
+                         * If stream doesn't have a rate, assume 1.0 (always
+                         * in-sample). Cache this determination for any future
+                         * use.
+                         */
+                        cache[stream] = true
+                    } else {
+                        /*
+                         * All platforms use session ID as the default
+                         * identifier for determining in- vs out-of-sample of
+                         * events sent to streams. On the web, streams can be
+                         * set to use pageview token instead. On the apps,
+                         * streams can be set to use device token instead.
+                         */
+                        var identifier = "session"
+                        if sampling_config.keys.contains("identifier") {
+                            identifier = sampling_config["identifier"] as! String
+                            if identifier != "session" && identifier != "device" {
+                                cache[stream] = false
+                                return false
+                            }
+                            identifier = identifier == "session" ? session_id() : device_id()
+                            let rate = sampling_config["rate"]! as! Double
+                            cache[stream] = determine(identifier, rate)
+                        } // end if there's identifier in sampling config
+                    } // end if there's rate in sampling config
+                } // end if there's a sampling config
+            } // end if there's a config
+        } // end if there's no cached determination
+        return cache[stream]!
```

Comment:
Here's a suggestion to avoid the extra nesting & `!`s:

```
        guard let config = config else {
            DDLogDebug("Invalid state, must have config to check for inSample")
            return false
        }
        
        if let cachedValue = cache[stream] {
            return cachedValue
        }

        guard let streamEntry = config[stream],
            let streamConfig = streamEntry["config"] as? [String: Any],
            let samplingConfig = streamConfig["sampling"] as? [String: Any] else {
                DDLogDebug("Failure pulling sampling key from config.")
                return false
        }
        
        /*
        * If cache does not have a determination for stream,
        * generate one.
        */
        guard let rate = samplingConfig["rate"] as? Double else {
            
            /*
            * If stream doesn't have a rate, assume 1.0 (always
            * in-sample). Cache this determination for any future
            * use.
            */
            cache[stream] = true
            return true
        }
        
        /*
        * All platforms use session ID as the default
        * identifier for determining in- vs out-of-sample of
        * events sent to streams. On the web, streams can be
        * set to use pageview token instead. On the apps,
        * streams can be set to use device token instead.
        */
        let identifierType = samplingConfig["identifier"] as? String ?? "session"
        
        guard identifierType == "session" || identifierType == "device" else {
            cache[stream] = false
            return false
        }
        
        let identifier = identifierType == "session" ? sessionID : storageManager.deviceID
        let result = determine(identifier, rate)
        cache[stream] = result
        return result
```

Reply:
I love this a lot because it's almost exactly like the JS client (https://gerrit.wikimedia.org/r/602739) and cross-platform similarity like that is cool and good.

P.S. Nuria was the one who pushed me away from nested conditionals in that patch.